drivers/adc: stm32: workaround for L462xE errata 2.5.2 and stuck reads

giansta opened the pull request:
According to the ST Errata sheet ES0389 - Rev 5, section 2.5.2, the
workaround to the issue of "Wrong ADC result if conversion done late
after calibration or previous conversion" is to perform two consecutive
ADC conversions.
We found also that the same objective is achieved enabling and disabling
it at every read. Moreover, this approach solves another issue found on
several STM32L462VE samples, where the reads were stuck, giving always
the same wrong value, that was 0 or the last calibration value.

Additional clearing of ADRDY flag has been added to avoid finding it
already set before next enabling.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

giansta committed:
drivers/adc: stm32: workaround for L462xE errata 2.5.2 and stuck reads

According to the ST Errata sheet ES0389 - Rev 5, section 2.5.2, the
workaround to the issue of "Wrong ADC result if conversion done late
after calibration or previous conversion" is to perform two consecutive
ADC conversions.
We found also that the same objective is achieved enabling and disabling
it at every read. Moreover, this approach solves another issue found on
several STM32L462VE samples, where the reads were stuck, giving always
the same wrong value, that was 0 or the last calibration value.

Additional clearing of ADRDY flag has been added to avoid finding it
already set before next enabling.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

giansta commented:
It looks like buildkite had errors on tests that don't seems to be related to this PR changes (e.g. tests/kernel/workq/work/kernel.work.api).
The "convert to draft - put in ready state" trick used in similar past cases doesn't help.

erwango commented:
I would like other ADC users could review

ABOSTM reviewed (approved):
@giansta, thanks for this PullRequest.
LGTM

str4t0m reviewed (changes requested):
I am not an extensive adc user, but would like to raise the following two concerns:

- After this change the ADC has to be enabled twice for a single read on G0 series, because of the errata that the ADC_cfgr1 can't be changed while the ADEN bit is set(needed for LL_ADC_SetResolution in start_read()).
   It would be great if this could be avoided.
- Am I right that the issue is present for the async api?

Could moving the disabling and enabling of the adc into the start_read() function solve both these problems, or do you see any problem that is introduced with such a move.

giansta commented:
@str4t0m, thanks for your suggestions. I tried to apply the changes you proposed.

I could test them with some limitation on our L462VE platform, as we're using an older branch without G0 family support, I'll try to test it more when possible. Unfortunately I don't have any G0 family board to test it.

giansta committed:
drivers/adc: stm32: workaround for both L462xE and G081xB errata

Apply workaround for both L462xE errata 2.5.2 and G081xB errata 2.6.2,
avoiding to have double enabling for a single read on G0 series.
Make workaround for L462xE valid also for the async api.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

cfriedt commented:
@giansta - is this one moving forward still?

cfriedt commented:
@str4t0m - can you please re-review?

cfriedt commented:
@anangl - could you also please review?

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -502,15 +502,61 @@ static void adc_stm32_isr(const struct device *dev)
+	} while (LL_ADC_IsActiveFlag_ADRDY(adc) == 0);
+	LL_ADC_ClearFlag_ADRDY(adc);
+#else
+	if (LL_ADC_IsEnabled(adc) == 0UL) {
```
This condition is redundant with the condition used to call this function.

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -502,15 +502,61 @@ static void adc_stm32_isr(const struct device *dev)
+	defined(CONFIG_SOC_SERIES_STM32H7X) || \
+	defined(CONFIG_SOC_SERIES_STM32WLX)
+	/*
+	 * Enabling ADC modules in L4, WB, G0 and G4 series may fail if they are
```
Comment is not up to date with the list of SoCs available if the #if .
Also, would you be able to point to errata for other series  ?
(it could be done in the commit message as for L4, that would be perfect)

str4t0m reviewed (changes requested):
I'm not sure if disabling of the ADC after each read should be configurable, but as applications which continuously sample adc values likely use a custom dma based driver doing it always should be fine. (The time it takes until the ADC is ready is equal to 1-2 conversion cycles on series I am familiar with.)

Assuming the ADC is disabled unconditionally after each read, special handling of stm32g0 in lines 412-425 is no longer needed.
However I'd prefer to add a short note to the driver about the special cases on stm32g0 and stm32l462xe. Otherwise these requirements are likely to be overlooked in future driver changes.

Additionally the ADC should also be in disabled state after `adc_stm32_init`, otherwise the behavior will  differ between the first and subsequent reads.
The same applies to `start_read`: In case calibration fails you should not return immediately, but also disable the ADC.

While the ADC is disabled unconditionally at the end of `start_read` it is never enabled for `CONFIG_SOC_SERIES_STM32F1X` and `STM32F3X_ADC_V2_5`. Could you please make sure that it is also enabled again on these series.

Could you eventually stash the two commits as the 2nd one reverts many changes of the first.

cfriedt commented:
@giansta , @str4t0m  - can we move this forward?

cfriedt commented:
@giansta - please address comments from @str4t0m
